### ISSUE TITLE
semver issues after upgrading to 5.4.1

### ISSUE THREAD

**Issue by fippo** [npm/npm#18434]
#### I'm opening this issue because:

  - [x] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.

#### What's going wrong?
```npm install github:opentok/rtcstats#semver:1.0.2```
is resulting in
```
npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for undefined
```
after upgrading to 5.4.1. Downgrading to 5.4.0 still works. Not sure if the #semver syntax I am using is correct even.

This from the npm log file looks relevant:
```
7 silly fetchPackageMetaData error for github:opentok/rtcstats#semver:1.0.2 No valid versions available for undefined
8 verbose type version
9 verbose stack Error: No valid versions available for undefined
9 verbose stack     at pickManifest (/usr/lib/node_modules/npm/node_modules/pacote/node_modules/npm-pick-manifest/index.js:19:11)
```

#### How can the CLI team reproduce the problem?
see above

### supporting information:

 - `npm -v` prints: 5.4.1
 - `node -v` prints: v6.9.5
- OS: Linux (ubuntu 16.04)

**Comment by ChrisRast** [comment]
Try removing the "semver" part and it works fine :
`npm install github:opentok/rtcstats#1.0.2`

**Comment by pushplay** [comment]
I'm experiencing the same or a smilar problem.

The first time I ran the install command I hadn't created the GitHub version yet so I got `No valid versions available for privateproject` which is reasonable.  Then I created the version and I got `No valid versions available for undefined`.  It worked after downgrading with `npm i -g npm@5.4.0`.

This is output from the run resulting in `No valid versions available for undefined`.

`npm i git+ssh://git@github.com/organization/privateproject.git#semver:^4.1.0`

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/Cellar/node/8.2.1/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'i',
1 verbose cli   'git+ssh://git@github.com/organization/privateproject#semver:^4.1.0' ]
2 info using npm@5.4.1
3 info using node@v8.2.1
4 verbose npm-session 4fa39b8959dd8759
5 silly install loadCurrentTree
6 silly install readLocalPackageData
7 silly fetchPackageMetaData error for git+ssh://git@github.com/organization/privateproject#semver:^4.1.0 No valid versions available for undefined
8 verbose type range
9 verbose stack Error: No valid versions available for undefined
9 verbose stack     at pickManifest (/usr/local/lib/node_modules/npm/node_modules/pacote/node_modules/npm-pick-manifest/index.js:19:11)
9 verbose stack     at git.revs.then.remoteRefs (/usr/local/lib/node_modules/npm/node_modules/pacote/lib/fetchers/git.js:136:7)
9 verbose stack     at tryCatcher (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/util.js:16:23)
9 verbose stack     at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:512:31)
9 verbose stack     at Promise._settlePromise (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:569:18)
9 verbose stack     at Promise._fulfillPromises (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:668:14)
9 verbose stack     at Promise._settlePromises (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:694:18)
9 verbose stack     at Promise._fulfill (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:638:18)
9 verbose stack     at Promise._resolveCallback (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:432:57)
9 verbose stack     at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:524:17)
9 verbose stack     at Promise._settlePromise (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:569:18)
9 verbose stack     at Promise._settlePromise0 (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:614:10)
9 verbose stack     at Promise._settlePromises (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:693:18)
9 verbose stack     at Promise._fulfill (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:638:18)
9 verbose stack     at Promise._resolveCallback (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:432:57)
9 verbose stack     at Promise._settlePromiseFromHandler (/usr/local/lib/node_modules/npm/node_modules/bluebird/js/release/promise.js:524:17)
10 verbose cwd /Users/jeff/github/organization-somerepo
11 verbose Darwin 16.7.0
12 verbose argv "/usr/local/Cellar/node/8.2.1/bin/node" "/usr/local/bin/npm" "i" "git+ssh://git@github.com/organization/privateproject#semver:^4.1.0"
13 verbose node v8.2.1
14 verbose npm  v5.4.1
15 error code ENOVERSIONS
16 error No valid versions available for undefined
17 verbose exit [ 1, true ]
```

**Comment by Alexsey** [comment]
This should fix the issue https://github.com/zkat/pacote/pull/119

**Comment by wlsf82** [comment]
I've just updated npm to version 5.4.2 and this still happens. The only way to avoid the issue so far is to pin on version 5.4.0.

**Comment by kohashi** [comment]
same problem. 

semver for git-repository is not working at version 5.4.2 and 5.4.1. 
all berrow commands, works fine at 5.4.0.

- `npm install github:myname/myrepo#semver:0.3.0`
- `npm install github:myname/myrepo#semver:~0.3.0`
- `npm install github:myname/myrepo#semver:~v0.3.0`
- `npm install github:myname/myrepo#semver:0.3.x`
- `npm install github:myname/myrepo#semver:v0.3.x`

**Comment by Alexsey** [comment]
This is because my fix is still not released. I really hope that it will get to the 5.4.3

**Comment by alexgig** [comment]
Can we get this fixed? Still an issue in 5.5.x. Thanks.

**Comment by kdelmonte** [comment]
@alexgig spoken like a true paying customer!

**Comment by alexgig** [comment]
Haha :-)

**Comment by Alexsey** [comment]
As far sa I can see this fix should already be in npm beta. You could try it if you need it right now

UPD: at list version range of pacote in npm beta dependencies should include the fixed version

UPD2: no this is not fixed because in package-lock even in master pacote is still 6.0.2, while fix is in 6.0.3. I think it needs a separate PR to bump the package-lock version. I can do it in a day so if any one has a time right now - It is really easy :)

**Comment by alexgig** [comment]
Hi @Alexsey,

Thank you for the updates. I've verified that it is still an issue in npm v5.5.1 which holds both `latest` and `next` tags.

Thank you for your help.

Best,
Alex

**Comment by fippo** [comment]
@Alexsey this was too tempting so I made a PR ;-)

**Comment by Alexsey** [comment]
@fippo You didn't change all the stuff in package-lock.json - try like this
```
"pacote": {
      "version": "6.0.3",
      "resolved": "https://registry.npmjs.org/pacote/-/pacote-6.0.3.tgz",
      "integrity": "sha512-VWEAIA6lwvlLQrChkiBWYWKpjx4HYG36ZHF1lVNWZyISZ8GcaVnMMUhi2MBTrbIDTn2VIAqH4OOEXXHaLR6Xkg==",
      "requires": {
        "bluebird": "3.5.0",
        "cacache": "9.2.9",
        "glob": "7.1.2",
        "lru-cache": "4.1.1",
        "make-fetch-happen": "2.5.0",
        "minimatch": "3.0.4",
        "mississippi": "1.3.0",
        "normalize-package-data": "2.4.0",
        "npm-package-arg": "5.1.2",
        "npm-packlist": "1.1.8",
        "npm-pick-manifest": "1.0.4",
        "osenv": "0.1.4",
        "promise-inflight": "1.0.1",
        "promise-retry": "1.1.1",
        "protoduck": "4.0.0",
        "safe-buffer": "5.1.1",
        "semver": "5.4.1",
        "ssri": "4.1.6",
        "tar": "4.0.1",
        "unique-filename": "1.1.0",
        "which": "1.3.0"
      }
    }
```

**Comment by Alexsey** [comment]
@zkat should PR propose 6.0.3, or 6.0.4 is better?

**Comment by alexgig** [comment]
I've verified that + pacote@6.0.4 works.

**Comment by alexgig** [comment]
Created a pr. Thx

**Comment by alexgig** [comment]
Hi @Alexsey,

How does the pr look? Can we get a quick patch release?

Thanks,
Alex

**Comment by Alexsey** [comment]
@alexgig As for me it looks good, but I'm not an npm maintainer. I think we should wait for comments on PR from npm team.  For now we can only 👍 the PR to show that there are people who are waiting on it

**Comment by alexgig** [comment]
awesome, thanks.

**Comment by zkat** [comment]
I'm working on dealing with other regressions in pacote related to git before the release, but the plan is to try and get a release out *hopefully* today.

**Comment by alexgig** [comment]
Hi @zkat,

Any updates on this? `panosoft/elm-grove` is being affected by this regression.

Thanks,
Alex

**Comment by alexgig** [comment]
Hi @zkat,

I don't mean to pester you on this but any update on progress? Unfortunately `panosoft/elm-grove` cannot be installed due to the pacote semver issue that `v6.0.4` resolves. Is there any way we can get a patch released that includes that? I know it may not be a fix of all regressions but at least it's a step forward.

Thanks for considering it.

Best,
Alex

**Comment by alexgig** [comment]
@zkat Can we get an update on this? Is there any way we can help?

**Comment by Alexsey** [comment]
@alexgig As I can see from zkat comment on this one https://github.com/npm/npm/pull/18896 it is already bumped. I have looked in to the branch and it is really 6.1.0 in there. So it is not about what we could do or about what zkat can do but about when would `release-next` would be released

**Comment by wilk** [comment]
Same problem here with private repos and semver:

- node 9.0.0
- npm 5.5.1

`$ npm i -S git+ssh://git@my.private.repo/repo/repo.git#semver:^1.2.0`

The error is:

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for undefined

**Comment by doesdev** [comment]
Man, this is a bit of a bummer. So I've been running 5.4.0 to sidestep this and thought I would update node then re-install 5.4.0. Turns out node 9 and 5.4.0 don't work together :/

Can't wait for a release

**Comment by EvgenyOrekhov** [comment]
@doesdev npm@5.4.0 has onother problem, sometimes it refuses to properly install dependencies saying

> Cannot read property 'pause' of undefined

In my experience the most stable of the recent versions is 5.3.0, and it works fine with Node.js 9.0.0 (though I use Docker and I have to install npm@5.3.0 using Yarn: `yarn global add npm@5.3.0`, you can try that).

**Comment by richardpringle** [comment]
Getting the same error on
npm 5.5.x
node 7.10.1 (although I don't think the node version makes a difference)

**Comment by den67rus** [comment]
In version **v5.6.0** everything works!
If you need to urgently upgrade:
```npm install npm@next -g```

**Comment by everdrone** [comment]
node `9.3.0`
npm `5.6.0`

```
npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for timed-out
```

**Comment by krichey15** [comment]
Node 8.9.4
Npm 5.6.0

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for timed-out

**Comment by Cassianosch** [comment]
Node 8.9.4
Npm 5.4.0

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for timed-out

**Comment by teawithfruit** [comment]
https://github.com/npm/registry/issues/255

**Comment by VarshiniSampath** [comment]
I'm getting the same error in Node v9.2.1 and npm v5.5.1. Is there a safe version of npm that can be used?

**Comment by nodediggity** [comment]
Node 8.9.4
Npm 5.4.0

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for timed-out

I've un/installed multiple versions via brew and no luck

**Comment by hacklock** [comment]
node 9.3.0
npm 5.6.0

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for timed-out

**Comment by guillenotfound** [comment]
node 9.3.0
npm 5.6.0

npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for timed-out

**Comment by miketimberg** [comment]
The `npm ERR! code ENOVERSIONS` occurs when building an empty Debian Jesse Docker image.

NodeJS: 8.9.4
npm: 5.6.0

**Comment by hacklock** [comment]
npm afk lol

**Comment by karthikiyengar** [comment]
Same thing.
npm: 5.6.0
node: 9.3.0

**Comment by teufernandesp** [comment]
Same here
node 8.9.4
npm 5.6.0

**Comment by karthikiyengar** [comment]
It's an issue with the registry apparently: https://status.npmjs.org/incidents/41zfb8qpvrdj

**Comment by teufernandesp** [comment]
I can't install Cordova  =/
aways show the message **The npm ERR! code ENOVERSIONS**

**Comment by teufernandesp** [comment]
I re-install node 8.9.0 and works for me!

**Comment by fippo** [comment]
the original issue seems to be resolved by now (tested with 5.6.0) so closing this

**Comment by naugtur** [comment]
Getting 
```
npm ERR! code ENOVERSIONS
npm ERR! No valid versions available for undefined
```
when installing dependencies from local git with #semver:  

node 6, npm 5.10.0
 ~~Same thing in npm 6.0.1~~ doesn't reproduce, might be a human error (meaning myself)

@fippo

**Comment by dmidz** [comment]
I finally get semver range to work by enclosing local network Git URL in quotes, checking npm logs showed carret were missing and so tried to install a specific commit version instead, ex :

`npm install "git+http://git.company.local/my-wonderfull-repo.git#semver:^1.0.9"`

will install last corresponding version such 1.0.13

using npm 5.8.0, node 6.11.2 (8.11.2 also) on Windows 10
